rfcs/rfc0010.md

```diff
@@ -0,0 +1,157 @@
+# A built-in for getting key-and-value pairs
+
+## Preamble
+
+    Author:  Ricardo Signes <rjbs@semiotic.systems>
+    Sponsor:
+    ID:      0010
+    Status:  Draft
+
+## Abstract
+
+This RFC proposes `kv`, a new function for returning the keys and values of a
+contianer.  Among other things, this allows key/value iteration on arrays as
+easily on hashes.
+
+## Motivation
+
+Right now, the `keys` and `values` functions operate on both arrays and hashes,
+giving the same semantics, if we accept that array indexes are effectively the
+keys of the array.  Pair slices, similarly, produce key/value pairs of array
+entries.  The sole missing component is that the behavior of a hash in list
+context is not easily reproduced.
+
+Providing this doesn't just fulfill a sense of completionism, but also has an
+immediate practical benefit.  With v5.36.0 poised to add n-at-a-time foreach,
+easily getting a list of index/value pairs from an array makes iteration over
+them easy.
+
+## Rationale
+
+If we start with the specific case of iterating over the indexes and values of
+an array using two-target foreach, we might write this:
+
+```perl
+for my ($i, $value) (%array[ keys @array ]) {
+  say "$i == $value";
+}
+```
+
+This is tolerable, but a bit verbose.  If we bury our target array deep in a
+structure, we get this:
+
+```perl
+for my ($i, $value) ($alpha->{beta}->[0]->%[ keys $alpha->{beta}->[0]->@* ]) {
+  say "$i == $value";
+}
+```
+
+This is pretty bad.
+
+With `kv`, we write this:
+
+```perl
+for my ($i, $value) (kv $alpha->{beta}->[0]->@*) {
+  say "$i == $value";
+}
+```
+
+This is probably about as simple as this can get without some significant new
+addition to the language.
+
+## Specification
+
+Like `keys` or `values`, `kv` can be called with a hash expression or array
+expression.  Other types of arguments are compile time errors.
+
+In scalar context, `kv` returns the number of entries in its argument, just
+like `keys` or `values`.
```

Comment:
I might have expected that it would return the size of the list it would return in list context, which would be twice the number of entries in its argument.

Also worth comparing to `->%[]` slices, which return the value of the last key/index requested.

Reply:
On the other hand list `kv %hash` is the same as `%hash` so making scalar kv match that seems simpler and easy to explain.  I think this is going to be easy to explain as:

* keys does X
* values does Y
* kv does Z

All three can be applied to hashes and arrays with clear and translatable meaning.  In list context, hashes act like kv was used and arrays like values.

Reply:
Slicing for "The whole thing" was proposed in another RFC, response was neutral-to-negative, and this proposal was meant to have far fewer ramifications, capable of getting into the language to make array iteration all better at once.

Reply:
I'm with @haarg here. I really would expect the invariant that:

```perl
(scalar kv %hash) == do { my @tmp = kv %hash; scalar @tmp }
```

I.e. that `scalar kv` returns twice the number of keys or values; the total number of keys and values combined.

Reply:
It is a good point that since `kv %hash` returns the same as `%hash` in list context, it would be reasonable for it to return the same in scalar context as well. But that also seems like optimizing for something that should never be used. `kv %hash` exists in this proposal for consistency, but is never meant to be used by anyone.

Reply:
Perhaps it would be better if it rejected hashes outright, so we can focus on its behavior for arrays. It needs a real array or hash, so there is no benefit to accepting both.